File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        
        sort(begin(nums), end(nums));
        vector<vector<int>> ans;
        int n=nums.size();
        
        for(int i = 0; i < n &&  nums[i] <= 0; i++){
            
            if(i==0 || nums[i-1]!=nums[i]){
                int l = i+1, r = n-1;
            
                while(l<r){
                    int sum = nums[i] + nums[l] + nums[r];
                
                    if(sum>0) r--;
                    else if(sum<0) l++;
                    else{
                        ans.push_back({nums[l++], nums[i], nums[r--]});
                        while(l<r && nums[l] ==  nums[l-1]) l++;
                    }
                }
            }
        }
        return ans;
    }
};

File: 0015-3sum/0015-3sum.java
class Solution {
    public List<List<Integer>> threeSum(int[] nums) {
        int n = nums.length;
        Arrays.sort(nums);

        List<List<Integer>> ans = new ArrayList<>();

        for(int i=0; i<n-1; i++){

            if(i>0 && nums[i] == nums[i-1]) continue;

            int left = i+1, right = n-1;
            
            while(left<right){
                int sum = nums[i] + nums[left] + nums[right];

                if(sum == 0){
                    ans.add(new ArrayList<>(Arrays.asList(nums[i], nums[left], nums[right])));

                    while(left < right && nums[left] == nums[left + 1]){
                        left++;
                    }

                    while(right > left && nums[right] == nums[right-1]){
                        right--;
                    }
                    left++;
                    right--;
                } else if(sum < 0){
                    left++;
                } else{
                    right--;
                }
            }

            
        }

        return ans;
    }
}